### DISCUSSION TITLE
Trying to understand how the IAudioPolicyConfigFactory interface works

### DISCUSSION THREAD

**Discussion by Chesewip** [File-New-Project/EarTrumpet#1543] [Q&A]
Hey yall, 

I'm working over in C++ land for an unrelated project, and I'm trying to just switch a processes output device through the winapi. I found this repo which seems to be one of the only apps that utilizes the undocumented IAudioPolicyConfigFactory interface. 

I have a pretty much identical setup from just converting things to C++, but when I actually call the functions like 
SetPersistedDefaultAudioEndpoint or the get method, I'm getting crashes or just no returns at all. (The GetPersistedDefaultAudioEndpoint returns an OK HRESULT but does not fill the value).

I can succesfully build the factory 

```cpp
       AudioPolicyConfigFactoryImplForDownlevel() {
            HRESULT hr = RoGetActivationFactory(Wrappers::HStringReference(L"Windows.Media.Internal.AudioPolicyConfig").Get(),
                __uuidof(IAudioPolicyConfigFactoryVariantForDownlevel),
                reinterpret_cast<void**>(m_factory.GetAddressOf()));
            if (FAILED(hr)) {
                throw std::runtime_error("Failed to get activation factory");
            }
        }
```
        

        So I know that I've atleast made contact with the COM object where ever this mysterious IAudioPolicyConfigFactory resides.
        But then going to the functions I get no results.
        
        Heres my definitions 

        
```cpp
        struct IAudioPolicyConfigFactory {
    virtual HRESULT SetPersistedDefaultAudioEndpoint(uint32_t processId, EDataFlow flow, ERole role, LPCWSTR deviceId) = 0;
    virtual HRESULT GetPersistedDefaultAudioEndpoint(uint32_t processId, EDataFlow flow, ERole role, HSTRING* deviceId) = 0;
    virtual HRESULT ClearAllPersistedApplicationDefaultEndpoints() = 0;
};
```

I'm pretty sure everything here matches the C# definitions ... but I'm having no results.
And I can run this at the same exact time I run the earTrumpet build, pass in the same PID and device ID, and no luck. 

 Any input or just information about this interface would be appreciated

**Comment by riverar** [comment]
@Chesewip Looks like you may have removed my [incomplete stubs in `IAudioPolicyConfigFactory`](https://github.com/File-New-Project/EarTrumpet/blob/master/EarTrumpet/Interop/MMDeviceAPI/IAudioPolicyConfigFactoryVariantFor21H2.cs#L10-L28)? These are required to ensure the virtual method table is lined up and compatible from an application binary interface perspective. In the snippet above, calling `ClearAllPersistedApplicationDefaultEndpoints` (method 3) would actually be invoking `add_RingerVibrateStateChanged` on the other end.

(You're free to name them whatever you want though!)

**Reply by Chesewip** [reply]
Oh my gosh thank you so much such a simple oversight on my part. You have no idea I've been beating my head against the wall trying to figure out the bug. COM is not my friend lmao

**Reply by riverar** [reply]
@Chesewip No problem, happy to help! (Also make sure you're using the IID from IAudioPolicyConfigFactoryImplFor21H2.) Any OS lower than 21H2 will need the Downlevel interface. (Microsoft mistakenly changed the IID between OS versions hence all this mess.)

**Reply by Chesewip** [reply]
Hey one more small question. 

I noticed in AudioDeviceManager theres this comment in the SetDefaultEndPointMethod
```        public void SetDefaultEndPoint(string id, int pid)
        {
            // Note: We found it unexpected that SetDefaultEndPoint
            // would accept an id for an 'invalid' device and then 
            // no audio will be heard from any device for the given pid.
            if (id == null || TryFind(id, out _))
            {
                _policyConfigService.SetDefaultEndPoint(id, pid);
            }
        }
```

Do you know if that is still a functionality? I was not able to reproduce this by passing in junk or null for id. 
I ask because I have a small niche case where sending off the process to nowhere would actually be useful.

**Reply by riverar** [reply]
@Chesewip Interesting hm, might have been bugfixed by the OS? Does it throw at all with a `null` passed in?

**Reply by Chesewip** [reply]
@riverar Sorry didn't see the reply, 

If you pass in null it doesn't throw, everything seems to work correctly, just the output seems to always default to the current system default. 

I've set the output device of the process before opening ear trumpet, open ear trumpet and pass in null, and then it goes back to system default.

**Reply by riverar** [reply]
@Chesewip Dang, sounds like they fixed it. Good for us, but bad for you! :(
